### DISCUSSION TITLE
How to define the Rendering location?

### DISCUSSION THREAD

**Discussion by Mark2Mark** [coldtype/coldtype#156] [Q&A] [answered]
I am currently trying to clean up my coldtype setup, and I think I don’t like to have all the scripts inside of the coldtype venv folder.

So my new approach is to have one or more coldtype folders with different versions as venvs. And then outside of those folders I got the "coldtype_scripts" folder, which will hold all my python files independently from the ct version (preferrably all script content adhering to the latest ct version though).

Anyway, for the sake of keeping things clean, I wonder if it is possible to also have a folder outside of everything where the frames and rendered videos can go, instead of next to the scripts into the "renders" folder. And if so, how to tell coldtype where to put those?

**Comment by stenson** [comment, marked as answer]
There's a command-line option (`--output-folder`/`-of`) that will let you completely override the render location, so you could do something like `$ coldtype sources/example.py -of override_renders/example` to have everything render to an arbitrary destination (here it's `override_renders/example`) — I think that's the easiest way to get what you're after, though that's not configurable in the code itself.

Another option would be to subclass the `animation` decorator, and reimplement the `pass_path` method, like so:

```python
from coldtype import *

class custom_output_animation(animation):
    def pass_path(self, index=0):
        return Path("custom_output_folder") / self.name / f"{self.pass_prefix()}{self.pass_suffix(index)}.{self.fmt}"

@custom_output_animation(bg=0)
def wght(f):
    return (StSt("CUSTOM\nOUTPUT", Font.MuSan(), 200, wght=f.e())
        .align(f.a.r)
        .f(1))
```

That’ll let you have control from inside the code on output, you'd just need to change any instances of `@animation` to `@custom_output_animation` (or whatever you'd like to call it)

Happy to chat about including it as a native concept in the `@animation` or `@renderable` constructors as well, though I have to admit I don't quite understand the use-case fully.  When you say "scripts inside the coldtype venv folder", do you mean adjacent to the venv folder? Or literally inside it? I don't ever put the coldtype source files inside the venv itself, so I'm not sure how the setup I use is different from the one you're describing.

**Reply by Mark2Mark** [reply]
Thanks. I will see how this works in a minute.

And to help you understand better: Oh yeah, sorry, I meant in the folder __next__ to the venv.

So what I had before:
```
Coldtype/
┣ animations/
┣ renders/
┣ venv/
```

but what I aim for is:

```
Coldtype-OLD/
┣ venv/

Coldtype-NEW/
┣ venv/

Coldtype_scripts/
┣ animations/
┣ other/

Coldtype_renderings/
┣ renders/

```

That way I can handle the sources better with git and have them not intertwined with the renderings. For the renderings, I usually want to version cotroll *some* but most of them I don’t. 

Having different Coldtype folders helps to switch back and forth when updating old scripts to new APIs (those could also be different branches, but also then, I don’t want the scripts and renderings inside of the Coldtype folders/branches)

**Reply by Mark2Mark** [reply]
Feedback on the `-of`: this works great for the single frames (`a` key in the viewer), but `r` releases the video into the parent directory of the defined location outpout option. Is that on purpoe?

Ah, sorry, my bad. I would need to define a subfolder, then it creates that inside the renderings folder, and the the video next to it. All good. This works quite nicely. I will see if I can make an alias for that, as it blows up the coldtype command a bit :)

**Reply by Mark2Mark** [reply]
Now I noticed, that when you use the -of option and the script has multiple animations, the frames get all thrown in the same folder (specified in the flag).

However, the same script run without the option renders the frames of the multiple animations in subfolders with their name.

Is there a way to get these subfolders also for the case when using this -of option?

**Reply by Mark2Mark** [reply]
Okay, after trying a lot of things and also shortcutting the -of option with a custom launch script, I currently find the option with the `@custom_output_animation` decorator working best for me.

I already have a litte custom library, so I added a `custom_settings` module. Now I just need to import that and change the decorator, and the files magically go to the desired location.

Really great, thanks Rob!

fyi, this is how I subclassed it now:
```python
class animation_custom_output(animation):
    def pass_path(self, index=0):
        parent_path = self.filepath.parent.name
        new_path = self.filepath.stem.replace(".coldtype", "")
        self.output_folder = Path(
            f"../coldtype_renders/{parent_path}/{new_path}/{self.name}"
        )
        if index is None:
            return self.output_folder / f"{self.pass_prefix()}"
        elif isinstance(index, int):
            return (
                self.output_folder
                / f"{self.pass_prefix()}{self.pass_suffix(index)}.{self.fmt}"
            )
        else:
            return self.output_folder / f"{self.pass_prefix()}{self.pass_suffix(index)}"
```

**Reply by stenson** [reply]
Great to hear it's working!
